onnxruntime/contrib_ops/webgpu/bert/group_query_attention.cc

```diff
@@ -110,35 +110,93 @@ Status GeneratePositionIDs(onnxruntime::webgpu::ComputeContext& context, const W
+                                    static_cast<int64_t>(head_size - half_rotary_embedding_dim)});
+  std::vector<uint32_t> k_global_dims(rank);
+  std::vector<uint32_t> k_global_strides(rank);
   for (size_t j = 0; j < rank; ++j) {
-    global_dims[j] = gsl::narrow_cast<uint32_t>(global_shape[j]);
-    global_strides[j] = gsl::narrow_cast<uint32_t>(global_shape.SizeFromDimension(j + 1));
+    k_global_dims[j] = gsl::narrow_cast<uint32_t>(k_global_shape[j]);
+    k_global_strides[j] = gsl::narrow_cast<uint32_t>(k_global_shape.SizeFromDimension(j + 1));
   }
-  const auto input_output_strides = std::vector<uint32_t>({gsl::narrow_cast<uint32_t>(input->Shape().SizeFromDimension(1)), gsl::narrow_cast<uint32_t>(hidden_size), gsl::narrow_cast<uint32_t>(head_size), 1});
-  const auto output_size = gsl::narrow_cast<const uint32_t>(global_shape.Size());
 
-  RotaryEmbeddingProgram program(params.rotary_interleaved_);
+  const auto q_domain_size = gsl::narrow_cast<uint32_t>(q_global_shape.Size());
+  const auto k_domain_size = gsl::narrow_cast<uint32_t>(k_global_shape.Size());
+
+  const auto q_input_output_strides = std::vector<uint32_t>(
+      {gsl::narrow_cast<uint32_t>(query_in->Shape().SizeFromDimension(1)),
+       gsl::narrow_cast<uint32_t>(hidden_size_q),
+       gsl::narrow_cast<uint32_t>(head_size),
+       1u});
+
+  const auto k_input_output_strides = std::vector<uint32_t>(
+      {gsl::narrow_cast<uint32_t>(key_in->Shape().SizeFromDimension(1)),
+       gsl::narrow_cast<uint32_t>(hidden_size_k),
+       gsl::narrow_cast<uint32_t>(head_size),
+       1u});
+
+  const uint32_t total = q_domain_size + k_domain_size;
+
+  FusedQKRotaryEmbeddingProgram program(params.rotary_interleaved_);
   program
       .CacheHint(params.rotary_interleaved_)
-      .AddInputs({{input, ProgramTensorMetadataDependency::Rank},
-                  {pos_ids, ProgramTensorMetadataDependency::Rank},
-                  {cos_cache, ProgramTensorMetadataDependency::Rank},
-                  {sin_cache, ProgramTensorMetadataDependency::Rank}})
-      .AddOutput(output)
-      .SetDispatchGroupSize((output_size + WORKGROUP_SIZE - 1) / WORKGROUP_SIZE)
-      .AddUniformVariables({{params.scale_},
-                            {gsl::make_span(global_dims)},
-                            {gsl::make_span(global_strides)},
-                            {gsl::make_span(input_output_strides)}})
+      .AddInputs({
+          {query_in, ProgramTensorMetadataDependency::Rank},
+          {key_in, ProgramTensorMetadataDependency::Rank},
+          {&pos_ids, ProgramTensorMetadataDependency::Rank},
+          {cos_cache, ProgramTensorMetadataDependency::Rank},
+          {sin_cache, ProgramTensorMetadataDependency::Rank},
+      })
+      .AddOutputs({
+          {query_out, ProgramTensorMetadataDependency::Rank},
+          {key_out, ProgramTensorMetadataDependency::Rank},
+      })
+      .SetDispatchGroupSize((total + WORKGROUP_SIZE - 1) / WORKGROUP_SIZE)
```

Comment:
Maybe we can optimize it by `total = q_domain_size` not `total = q_domain_size + k_domain_size`. And get the updated Q and K in one pass not one after another in shader since Q has shape [batch_size, sequence_length, num_head, head_size] and K has shape [batch_size, sequence_length, num_head_k, head_size] and num_head_k is always less than or equal to num_head.